pkg/agent/plugin/nodeattestor/aws/iid.go

```diff
@@ -20,52 +22,41 @@ import (
 )
 
 const (
-	iidPluginName               = "aws_iid"
-	defaultIdentityDocumentUrl  = "http://169.254.169.254/latest/dynamic/instance-identity/document"
-	defaultIdentitySignatureUrl = "http://169.254.169.254/latest/dynamic/instance-identity/signature"
+	defaultIdentityDocumentHost  = "169.254.169.254"
+	defaultIdentityDocumentPath  = "/latest/dynamic/instance-identity/document"
+	defaultIdentitySignatureHost = "169.254.169.254"
+	defaultIdentitySignaturePath = "/latest/dynamic/instance-identity/signature"
 )
 
+// IIDAttestorConfig configures a IIDAttestorPlugin.
 type IIDAttestorConfig struct {
-	IdentityDocumentUrl  string `hcl:"identity_document_url"`
-	IdentitySignatureUrl string `hcl:"identity_signature_url"`
+	IdentityDocumentURL  string `hcl:"identity_document_url"`
+	IdentitySignatureURL string `hcl:"identity_signature_url"`
+	AgentPathTemplate    string `hcl:"agent_path_template"`
+	trustDomain          string
+	pathTemplate         *template.Template
 }
 
+// IIDAttestorPlugin implements aws nodeattestation in the agent.
 type IIDAttestorPlugin struct {
-	trustDomain          string
-	identityDocumentUrl  string
-	identitySignatureUrl string
-
-	mtx *sync.RWMutex
+	config *IIDAttestorConfig
+	mtx    sync.RWMutex
 }
 
-func (p *IIDAttestorPlugin) spiffeID(awsAccountId, awsRegion, awsInstanceId string) *url.URL {
-	spiffePath := path.Join("spire", "agent", iidPluginName, awsAccountId, awsRegion, awsInstanceId)
-	id := &url.URL{
-		Scheme: "spiffe",
-		Host:   p.trustDomain,
-		Path:   spiffePath,
-	}
-	return id
+// NewIIDPlugin creates a new IIDAttestorPlugin.
+func NewIIDPlugin() *IIDAttestorPlugin {
+	return &IIDAttestorPlugin{}
 }
 
-func httpGetBytes(url string) ([]byte, error) {
-	resp, err := http.Get(url)
-	if err != nil {
-		return nil, err
-	}
-	defer resp.Body.Close()
-	bytes, err := ioutil.ReadAll(resp.Body)
+// FetchAttestationData fetches attestation data from the aws metadata server and sends an attestation response
+// on given stream.
+func (p *IIDAttestorPlugin) FetchAttestationData(stream nodeattestor.FetchAttestationData_PluginStream) error {
+	c, err := p.getConfig()
 	if err != nil {
-		return nil, err
+		return err
 	}
-	return bytes, nil
-}
 
-func (p *IIDAttestorPlugin) FetchAttestationData(stream nodeattestor.FetchAttestationData_PluginStream) error {
-	p.mtx.RLock()
-	defer p.mtx.RUnlock()
-
-	docBytes, err := httpGetBytes(p.identityDocumentUrl)
+	docBytes, err := httpGetBytes(identityURL(c.IdentityDocumentURL))
```

Comment:
This breaks backwards compatibility. Existing configuration that is setting these values is setting the whole url (e.g. http://host/path). `identityURL` does not handle this gracefully and would end up producing a URL `http://http:/host/path`, which is malformed. We should not break compatibility lightly.

I'm still not seeing the benefit for the host/path split in this PR. Unless there is a compelling reason for the change, I'd be more comfortable with that portion of this PR being reverted.

If there is something I'm missing, please let me know. I'm still open to being convinced. In any case, we should do whatever we can to prevent a breaking change.

Reply:
reverting and making constants for test file